libraries/AC_Avoidance/AP_OADijkstra.cpp

```diff
@@ -0,0 +1,550 @@
+            return false;
+        }
+    }
+
+    // rebuild path if destination has changed
+    if (!destination.same_latlon_as(_destination_prev)) {
+        _destination_prev = destination;
+        _shortest_path_ok = false;
+    }
+
+    // calculate shortest path from current_loc to destination
+    if (!_shortest_path_ok) {
+        _shortest_path_ok = calc_shortest_path(current_loc, destination);
+        if (!_shortest_path_ok) {
+            return false;
+        }
+        // start from 2nd point on path (first is the original origin)
+        _path_idx_returned = 1;
+    }
+
+    // path has been created, return latest point
+    Vector2f dest_pos;
+    if (get_shortest_path_point(_path_idx_returned, dest_pos)) {
+
+        // for the first point return origin as current_loc
+        Vector2f origin_pos;
+        if ((_path_idx_returned > 0) && get_shortest_path_point(_path_idx_returned-1, origin_pos)) {
+            // convert offset from ekf origin to Location
+            Location temp_loc(Vector3f(origin_pos.x, origin_pos.y, 0.0f));
+            origin_new = temp_loc;
+        } else {
+            // for first point use current loc as origin
+            origin_new = current_loc;
+        }
+
+        // convert offset from ekf origin to Location
+        Location temp_loc(Vector3f(dest_pos.x, dest_pos.y, 0.0f));
+        destination_new = destination;
+        destination_new.lat = temp_loc.lat;
+        destination_new.lng = temp_loc.lng;
+
+        // check if we should advance to next point for next iteration
+        const bool near_oa_wp = current_loc.get_distance(destination_new) <= 2.0f;
+        const bool past_oa_wp = current_loc.past_interval_finish_line(origin_new, destination_new);
+        if (near_oa_wp || past_oa_wp) {
+            _path_idx_returned++;
+        }
+        return true;
+    }
+
+    return false;
+}
+
+// check if polygon fence has been updated since we created the inner fence. returns true if changed
+bool AP_OADijkstra::check_polygon_fence_updated() const
+{
+    // exit immediately if polygon fence is not enabled
+    const AC_Fence *fence_ptr = AC_Fence::get_singleton();
+    if (fence_ptr == nullptr) {
+        return false;
+    }
+    return (_polyfence_update_ms != fence_ptr->get_boundary_update_ms());
+}
+
+// create a smaller polygon fence within the existing polygon fence
+// returns true on success
+bool AP_OADijkstra::create_polygon_fence_with_margin(float margin_cm)
+{
+    // exit immediately if polygon fence is not enabled
+    const AC_Fence *fence_ptr = AC_Fence::get_singleton();
+    if (fence_ptr == nullptr) {
+        return false;
+    }
+
+    // get polygon boundary
+    uint16_t num_points;
+    const Vector2f* boundary = fence_ptr->get_boundary_points(num_points);
+    if ((boundary == nullptr) || (num_points < 3)) {
+        return false;
+    }
+
+    // fail if fence is too large
+    if (num_points > OA_DIJKSTRA_POLYGON_FENCE_PTS) {
```

Comment:
Pretty sure this need to be `>=` because you copy the first point to be the last one, which means you can only accept 1 fewer point into `_polyfence_pts`

Reply:
ah, thanks.  with Peter's recent changes we don't need to close the polygon anymore so I'll remove that bit.